src/lando/main/scm/exceptions.py

```diff
@@ -0,0 +1,80 @@
+from typing import Optional
+
+from lando.api.legacy.hg import (
+    AutoformattingException as LegacyAutoformattingException,
+)
+from lando.api.legacy.hg import (
+    HgmoInternalServerError,
+    LostPushRace,
+    NoDiffStartLine,
+    PatchConflict,
+    PushTimeoutException,
+)
+from lando.api.legacy.hg import (
+    TreeApprovalRequired as LegacyTreeApprovalRequired,
+)
+from lando.api.legacy.hg import (
+    TreeClosed as LegacyTreeClosed,
+)
+
+
+class ScmException(Exception):
```

Comment:
I'm not sure I follow what the `ScmException` and this re-mapping functionality is achieving, as the custom exceptions should be sufficient for both Git and Hg. With this change it seems we throw our own custom exception when we hit known conditions, then re-map it to these new custom exception types later on, which are themselves subclassed on another custom exception.

Most of the existing exceptions are thrown for things like "we detected that this push failed due to a `LostPushRace`, throw this custom exception so it is caught in the landing worker and the landing job status can be set appropriately". If we need to abstract those details at the SCM level, perhaps we want to make the existing exceptions subclass our new type instead of doing the conversion on the fly.

Reply:
The ScmException should replace the Hg-only exceptions. I didn't want to change the legacy code, nor to make it depend on non-legacy additions, which is why we have this remapping dance for exceptions that are essentially exactly the same, except the new ones are in a more practical namespace to cater for multiple SCMs.